### PR TITLE
refactor(iroh-net): Proper abstraction around `best_addr`

### PR DESCRIPTION
**Pull request by Frando**
## Description

This refactors how we use and calculate `best_addr`. It should not have any functional changes, only refactoring to make things more robust for the future.

## Change checklist

- [x] Self-review.
- [ ] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (dignifiedquire) [inline iroh-net/src/magicsock/endpoint/best_addr.rs:48]: just `ClearReason`,  to avoid `best_addr::BestAddrClearReson` double

Comment (Frando) [reply iroh-net/src/magicsock/endpoint/best_addr.rs]: fixed!

**Review by dignifiedquire — APPROVED** [review]
lgtm but should get review from @divagant-martian before merging

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint/best_addr.rs:1]: this file is not pub but it would be good to document its contents for us

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint/best_addr.rs:11]: These two only exist for the best address, maybe move them to this file? Having a separate file is a really good call, helping cleanup `endpoint.rs` and my moving we can further tidy it

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint/best_addr.rs:69]:
```suggestion
    }
    
```

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint/best_addr.rs:168]: since the best address is per peer, these logs are probably more useful in the caller (where you already have similar logs)

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:210]: can you add `self.public_key.fmt_short()` here to know what peer we are working on?

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:215]: same here

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:225]: here as well

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:234]: this return value is never used in the code, we can remove it

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:153]: to me it makes sense what you say. When it's outdated we return both this address and the derp and metrics are only updated when the address is cleared. So even there, we are counting an outdated conn as a a direct one

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [inline iroh-net/src/magicsock/endpoint.rs:668]: this is a lot nicer but we lost the information of the peer whose best address is being dropped

**Review by divagant-martian — COMMENTED** [review]
overall I think it's a great improvement, and moving it to another file seems like the right call to me. I found a bug -not caused by your changes- but that we should address, and otherwise just some comments about info we lost in logging

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:210]: These log lines are only called from function that have instrumentation spans over them. So the log lines should look like this:
```
2023-10-23T09:20:32.398115Z TRACE poll_send{me=pp6ebwzioaeldcnl}:get_send_addrs{peer=cx6rezjzqm3ayple}: iroh_net::magicsock::endpoint: best_addr is set and valid, use best_addr only addr=172.20.0.1:60955 latency=Some(17.247288ms)
```

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:255]: Ah, true that `SendAddr::as_socket_addr` maps Derp addresses to fake addresses. However, in this place I think it is correct (even though not enforced on the type level), because we only use pongs and thus `SendAddr`s out of `Endpoint::direct_addr_state`, which should only ever contain pongs for direct (UDP) addresses. So I do think that this is correct here, unless I'm missing something.

I already thought of abstracing the direct addr state in a similar manner to this PR.

In any case - if the behavior is the same as it is on `main`, let's leave this to a followup and verify there that no fake Derp addresses can ever end up in the `best_addr`.

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:237]: Renamed to `assign_best_addr_from_candidates_if_empty`

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:275]: :+1: Changed to match on `best_addr.state()`in the latest commit, reads much clearer also.

File: iroh-net/src/magicsock/endpoint/best_addr.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint/best_addr.rs:168]: yup removed them here

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:234]: removed

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs]: I changed it so that we return `Direct` if best_addr is set but expired.

Actually the more proper fix might be to return something like `ConnectionType::Both` because if the `best_addr` is set but expired, we send over both Derp and the Udp address. But not sure how important this is to report to the outside.

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:217]: see above

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs:227]: see above

File: iroh-net/src/magicsock/endpoint.rs
Comment (Frando) [reply iroh-net/src/magicsock/endpoint.rs]: In the most recent commit I changed this to return `ConnectionType::Mixed`  in that case. Also put it all into a single `match` statement for clarity.

File: iroh-net/src/magicsock/endpoint.rs
Comment (divagant-martian) [reply iroh-net/src/magicsock/endpoint.rs:262]: ah yes, I fixed that in fact. Probably the reason why it's still there is because we store a pong for the derper (in the derp_region) as well

**Review by divagant-martian — APPROVED** [review]

**Pull request merged**
